devito/data/utils.py

```diff
@@ -0,0 +1,259 @@
+    idx = decomposition.index_glb_to_loc(idx, rel=False)
+
+    if is_integer(idx):
+        return PROJECTED
+    elif idx is None:
+        return NONLOCAL
+    elif isinstance(idx, (tuple, list)):
+        return [i - value for i in idx]
+    elif isinstance(idx, np.ndarray):
+        return idx - value
+    elif isinstance(idx, slice):
+        if idx.step is not None and idx.step < 0:
+            if idx.stop is None:
+                return slice(idx.start - value, None, idx.step)
+        return slice(idx.start - value, idx.stop - value, idx.step)
+    else:
+        raise ValueError("Cannot apply shift to type `%s`" % type(idx))
+
+
+def convert_index(idx, decomposition, mode='glb_to_loc'):
+    """Convert a global index into a local index or vise versa according to mode."""
+    if is_integer(idx) or isinstance(idx, slice):
+        return decomposition(idx, mode=mode)
+    elif isinstance(idx, (tuple, list)):
+        return [decomposition(i, mode=mode) for i in idx]
+    elif isinstance(idx, np.ndarray):
+        return np.vectorize(lambda i: decomposition(i, mode=mode))(idx)
+    else:
+        raise ValueError("Cannot convert index of type `%s` " % type(idx))
+
+
+def index_handle_oob(idx):
+    # distributed.index_glb_to_loc returns None when the index is globally
+    # legal, but out-of-bounds for the calling MPI rank
+    if idx is None:
+        return NONLOCAL
+    elif isinstance(idx, (tuple, list)):
+        return [i for i in idx if i is not None]
+    elif isinstance(idx, np.ndarray):
+        if idx.dtype == np.bool_:
+            # A boolean mask, nothing to do
+            return idx
+        elif idx.ndim == 1:
+            return np.delete(idx, np.where(idx == None))  # noqa
+        else:
+            raise ValueError("Cannot identify OOB accesses when using "
+                             "multidimensional index arrays")
+    else:
+        return idx
+
+
+def local_data_idx(loc_idx):
+    """Retrieve the unflipped idx corresponding to loc_idx."""
+    loc_data_idx = []
+    for i in as_tuple(loc_idx):
+        if isinstance(i, slice) and i.step is not None and i.step == -1:
+            if i.stop is None:
+                loc_data_idx.append(slice(0, i.start+1, -i.step))
+            else:
+                loc_data_idx.append(slice(i.stop+1, i.start+1, -i.step))
+        elif isinstance(i, slice) and i.step is not None and i.step < -1:
+            if i.stop is None:
+                lmin = i.start
+                while lmin >= 0:
+                    lmin += i.step
+                loc_data_idx.append(slice(lmin-i.step, i.start+1, -i.step))
+            else:
+                loc_data_idx.append(slice(i.stop+1, i.start+1, -i.step))
+        elif is_integer(i):
+            loc_data_idx.append(slice(i, i+1, 1))
+        else:
+            loc_data_idx.append(i)
+    return as_tuple(loc_data_idx)
+
+
+def mpi_index_maps(loc_idx, shape, topology, coords, comm):
+    """Generate the various mpi index maps."""
```

Comment:
what is an index map? I could use a proper example here

Reply:
Will elaborate. But basically some structures are created which tell us to which index `(i, j, k, ...)` should go i.e. `(i, j, k, ...) -> (I, J, K, ...)`.